dev-tools/release-notes/main/main.go

```diff
@@ -0,0 +1,266 @@
+	"strconv"
+	"strings"
+)
+
+const (
+	baseURL             = "https://api.github.com/repos/"
+	repo                = "elastic/cloud-on-k8s/"
+	releaseNoteTemplate = `:issue: https://github.com/{{.Repo}}issues/
+:pull: https://github.com/{{.Repo}}pull/
+
+[[release-notes-{{.Version}}]]
+== {n} version {{.Version}}
+{{range $group, $prs := .Groups}}
+[[{{- id $group -}}-{{$.Version}}]]
+[float]
+=== {{index $.GroupLabels $group}}
+{{range $prs}}
+* {{.Title}} {pull}{{.Number}}[#{{.Number}}]{{with .RelatedIssues -}}
+{{$length := len .}} (issue{{if gt $length 1}}s{{end}}: {{range $idx, $el := .}}{{if $idx}}, {{end}}{issue}{{$el}}[#{{$el}}]{{end}})
+{{- end}}
+{{- end}}
+{{end}}
+`
+)
+
+var (
+	groupLabels = map[string]string{
+		">breaking":    "Breaking changes",
+		">deprecation": "Deprecations",
+		">feature":     "New features",
+		">enhancement": "Enhancements",
+		">bug":         "Bug fixes",
+		"nogroup":      "Misc",
+	}
+
+	ignore = map[string]bool{
+		">non-issue":   true,
+		">refactoring": true,
+		">docs":        true,
+		">test":        true,
+		":ci":          true,
+		"backport":     true,
+	}
+)
+
+// Label models a subset of a GitHub label.
+type Label struct {
+	Name string `json:"name"`
+}
+
+// Issue models a subset of a Github issue.
+type Issue struct {
+	Labels        []Label           `json:"labels"`
+	Body          string            `json:"body"`
+	Title         string            `json:"title"`
+	Number        int               `json:"number"`
+	PullRequest   map[string]string `json:"pull_request,omitempty"`
+	RelatedIssues []int
+}
+
+type GroupedIssues = map[string][]Issue
+
+type TemplateParams struct {
+	Version     string
+	Repo        string
+	GroupLabels map[string]string
+	Groups      GroupedIssues
+}
+
+func fetch(url string, out interface{}) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+
+	nextLink := extractNextLink(resp.Header)
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return "", errors.New(fmt.Sprintf("%s: %d %s ", url, resp.StatusCode, resp.Status))
+	}
+
+	defer resp.Body.Close()
+	if err = json.NewDecoder(resp.Body).Decode(&out); err != nil {
+		return "", err
+	}
+	return nextLink, nil
+}
+
+func extractNextLink(headers http.Header) string {
+	var nextLink string
+	nextRe := regexp.MustCompile(`<([^>]+)>; rel="next"`)
+	links := headers["Link"]
+	for _, lnk := range links {
+		matches := nextRe.FindAllStringSubmatch(lnk, 1)
+		if matches != nil && matches[0][1] != "" {
+			nextLink = matches[0][1]
+			break
+		}
+	}
+	return nextLink
+}
+
+func fetchVersionLabels() ([]string, error) {
+	var versionLabels []string
+	url := fmt.Sprintf("%s%slabels?page=1", baseURL, repo)
+FETCH:
+	var labels []Label
+	next, err := fetch(url, &labels)
+	if err != nil {
+		return nil, err
+	}
+	for _, l := range labels {
+		if strings.HasPrefix(l.Name, "v") {
+			versionLabels = append(versionLabels, l.Name)
+		}
+	}
+	if next != "" {
+		url = next
+		goto FETCH
```

Comment:
good to have this so it's easier to rewrite in assembly language when we want more performance 😄 (kidding, `goto` are rather unusual, but fine with me)